Improve JWT Assertion Validation using client validators

rmartinc opened the pull request:
Closes #43642

The PR creates a new intermediate class `AbstractJWTValidator` with part of the methods that previously were in `AbstractJWTClientValidator`. Class `JWTAuthorizationGrantValidationContext` in `server-spi-private` is now an interface (very limited now, but we can add mode methods if needed) and the real code has been moved to `JWTAuthorizationJWTValidator` in `services` which extends the `AbstractJWTValidator` to inherit the old methods. This way the code is shared. The signature validation is performed in the IdP side, and we can move more parts to the IdP if we want. So the current code is something in the middle. The token is validated using the client validator, but the signature is checked by the IdP. Draft for the moment. Questions:

* Do you see this OK? Or do you prefer move signature to the client validator part? Or vice-versa, move token validations to the IdP side? It's a bit repeated in both sides.
* The `nbf` claim is not checked now, I have just detected it. The client validator was not doing it before, so I will add later.
* I have added the options we have currently in the IdP side (clockSkew and algorithm), but there are other options that we can add: token lifespan (default is 5m=300s) and reusePermitted (default to false). Do we want them or is it OK using fixed values for now?

I need to polish this a bit but the I want to hear your comments about the three questions before. :smile:

rmartinc committed:
Improve JWT Assertion Validation using client validators
Closes #43642

Signed-off-by: rmartinc <rmartinc@redhat.com>

mposolda reviewed (commented):
@rmartinc Cool, Thanks!

I have few inline comments (Some of them are more a questions from discussion. Feedback welcome from you and @graziang as well).

Some answers to your questions:


>    Do you see this OK? Or do you prefer move signature to the client validator part? Or vice-versa, move token validations to the IdP side? It's a bit repeated in both sides.

I see this OK as it is now in your PR.

>    The nbf claim is not checked now, I have just detected it. The client validator was not doing it before, so I will add later.

In `AbstractBaseJWTValidator.validateTokenActive` there is a call to `token.isActive` , which checks also "not before" . Is it sufficient or do we need something more?

>    I have added the options we have currently in the IdP side (clockSkew and algorithm), but there are other options that we can add: token lifespan (default is 5m=300s) and reusePermitted (default to false). Do we want them or is it OK using fixed values for now?

IMO ok to hardcode "reusePermitted" to false for now.  There is the other PR https://github.com/keycloak/keycloak/pull/43841/files , which introduces this including proper config.

For the "token lifespan", I am not sure which token you mean? Lifespan of the assertion is handled by the attributes of JWT itself. Lifespan of the issued token is the same for "jwt-bearer" grant like for any other grants AFAIK.

rmartinc commented on services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java:
```diff
@@ -1079,4 +1084,13 @@ public BrokeredIdentityContext validateAuthorizationGrantAssertion(JWTAuthorizat
+
+    @Override
+    public String getClientAssertionSigningAlg() {
+        return getConfig().getClientAssertionSigningAlg();
```
Yes, this was not even added by @graziang, I just added because it was in the base class and makes some sense. We can remove it for the moment or change the tooltip and doc to mention that this is the algorithm to sign any client or authorization grant. Whatever you prefer is OK to me.

rmartinc commented:
I have removed the TODO in the `OIDCIdentityProvider` class and removed the `getClientAssertionSigningAlg` in the `JWTAuthorizationGrantProvider`. With this the only remaining question is if we add more configuration methods in the JWT Authorization Grant:

* The signing algorithm for the JWT authorization grant (same prop or new).
* If reusing the same token is permitted.
* The maximum expiration time that is allowed for the grant.

We can manage those as a follow-up if we agree that any of those options are needed too.

mposolda commented:
@rmartinc Could you please rebase? I think that should help with the failing UI tests (not 100% sure, but I guess it might be related to https://github.com/keycloak/keycloak/commit/0f01444543cf5e755b13a62c233a5c9feef72888 )
